typing/rec_check.ml

```diff
@@ -20,144 +20,160 @@ exception Illegal_expr
+        For example, the subterm [?] is in a [Guard] context
+        in [let _ = ? in e] and in [?; e].
+        When these subexpressions participate in a cyclic definition,
+        they cannot create a self-loop.
+    *)
+
+    | Delay
+    (** A [Delay] context can be fully evaluated without evaluting its argument,
+        which will only be needed at a later point of program execution. For
+        example, [fun x -> ?] or [lazy ?] are [Delay] contexts. *)
+
+    | Unused
+    (** [Unused] is for subexpressions that are not used at all during
+       the evaluation of the whole program. This is the mode of
+       a variable in an expression in which it does not occur. *)
+
+  (* Returns the most conservative mode of the two arguments.
+
+     Deref < Return < Guard < Delay < Unused
+  *)
+  let prec m m' =
+    match m, m' with
+    | Dereference, _
+    | _, Dereference -> Dereference
+    | Return, _
+    | _, Return -> Return
+    | Guard, _
+    | _, Guard -> Guard
+    | Delay, _
+    | _, Delay -> Delay
+    | _ -> Unused
+
+  (* If x is used with the mode m in e[x], and e[x] is used with mode
+     m' in e'[e[x]], then x is used with mode m'[m] in e'[e[x]].
+
+     Return is neutral for composition: m[Return] = m = Return[m].
+  *)
+  let compos m' m = match m', m with
+    | Unused, _
+    | _, Unused -> Unused
+    | Dereference, _ -> Dereference
+    | Delay, _ -> Delay
+    | m', Return -> m'
+    | Return, m
+    | Guard, m -> m
+
+  module Env :
   sig
     type t
-    val guard : t -> t
-    (** An expression appears in a guarded context *)
 
-    val discard : t -> t
-    (** The address of a subexpression is not used, but may be bound *)
+    val single : Ident.t -> mode -> t
+    (** Create an environment with a single identifier used with a given mode.
+    *)
 
-    val inspect : t -> t
-    (** The value of a subexpression is inspected with match, application,
-        etc. *)
+    val empty : t
+    (** An environment with no used identifiers. *)
 
-    val delay : t -> t
-    (** An expression appears under 'fun p ->' or 'lazy' *)
+    val find : Ident.t -> t -> mode
+    (** Find the mode of an indentifier in an environment.  The default mode is
+        Unused. *)
 
-    val join : t -> t -> t
-    (** Combine the access information of two expressions *)
+    val unguarded : t -> Ident.t list -> Ident.t list
+    (** unguarded e l: the list of all identifiers in e that are unguarded of
+        dereferenced in the environment e. *)
 
-    val single : Ident.t -> access -> t
-    (** Combine the access information of two expressions *)
+    val dependent : t -> Ident.t list -> Ident.t list
+    (** unguarded e l: the list of all identifiers in e that are used in e. *)
 
-    val empty : t
-    (** No variables are accessed in an expression; it might be a
-        constant or a global identifier *)
+    val join : t -> t -> t
+    val join_list : t list -> t
```

Comment:
It would maybe nice to have a comment indicating that environments sharing bindings is supported, and that in that case the right thing happens (the most conservative mode is picked).